doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+      fixing the macro)
+
+   * The UFunc machinery changes will break *limited* parts of the current
+     implementation. Replacing e.g. the default ``TypeResolver`` is expected
+     to remain supported for a time, although optimized masked inner loop iteration
+     (which is not even used *within* NumPy) is expected to not remain supported
+     and lead to errors instead.
+
+   * All current functions defined on the dtypes currently, such as
+     ``PyArray_Descr->f->nonzero``, will move to a new home and recieve new
+     ways to access them.
+     This means that in the long run some (rare) lowlevel access code will
+     have to be changed to use the new API.
+
+* **dtype implementors (C-API)**:
+
+  * The array that is currently provided to some functions (such as cast functions),
+    may not be provided anymore generally.
+    For example ``PyArray_Descr->f->nonzero`` or ``PyArray_Descr->f->copyswapn``,
+    may instead receive a dummy array object with only some fields (mainly the
+    dtype), being valid.
+    At least in some code paths, a similar mechanism is already used.
+
+  * The ``scalarkind`` slot and registration of scalar casting will be
+     removed/ignored without replacement.
+     It currently allows partial value based casting.
+     The ``PyArray_ScalarKind`` function will continue to work for builtin types,
+     but will not be used internally and be deprecated.
+
+   * Currently user dtypes are defined as instances of ``np.dtype``.
+     The creation works by the user providing a prototype instance.
+     NumPy will need to modify at least the type during registration.
+     This has no effect for either ``rational`` or ``quaternion`` and mutation
+     of the structure seems unlikely after registration.
+
+Since there is a fairly large API surface concerning datatypes, further changes
+or the limitation of a certain function to currently existing datatypes is
+likely to occur.
+For example functions which use the type number as input
+should be replaced with functions taking DType classes instead.
+Although public, large parts of this C-API seems very rarely and possibly
+never used by downstream projects.
+
+
+
+
+Detailed Description
+--------------------
+
+This section details some of the design decisions above and gives
+more details.
+Since datatype changes touch a large part of code and behaviours, NEP 40
+reviews some of the concepts, issues, and the current implementation.
+
+
+Datatypes as Python Classes (1)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The current NumPy datatypes are not full scale python classes.
+They are instead (prototype) instances of a single ``np.dtype`` class.
+Changing this means that any special handling, e.g. for ``datetime``
+can be moved to the Datetime DType class instead, away from monolithic general
+code (e.g. current ``PyArray_AdjustFlexibleDType``).
+
+The main API side effect of this is that special methods move
```

Comment:
The main consequence of this change with respect to the API is that special methods move from the dtype instances to methods on the new dtype class.